### PR TITLE
[Go] Set Default Values for Required Variables where only one value is possible

### PR DESCRIPTION
**Pull request by prmedipa**
Fix [19107](https://github.com/OpenAPITools/openapi-generator/issues/19107)

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by wing328** [comment]
Thanks for the PR. I dont think required fields or parameters should have default values. Agreed?

**Comment by wing328** [comment]
Thanks for the PR. I dont think required fields or parameters should have default values. Agreed?

**Comment by vvb** [comment]
Ideally yes, but this is just a fallback in cases when only one value is possible as per the OAS. The Python SDK also has the same behaviour implemented.

File: docs/generators/fsharp-functions.md [hunk 1]
```diff
@@ -34,7 +34,7 @@ These options may be applied as additional-properties (cli) or configOptions (pl
 |prependFormOrBodyParameters|Add form or body parameters to the beginning of the parameter list.| |false|
 |sortModelPropertiesByRequiredFlag|Sort model properties to place required parameters before optional parameters.| |true|
 |sortParamsByRequiredFlag|Sort method arguments to place required parameters before optional parameters.| |true|
-|sourceFolder|source folder for generated code| |OpenAPI/src|
+|sourceFolder|source folder for generated code| |OpenAPI\src|
```
Comment (wing328) [inline docs/generators/fsharp-functions.md:37]: samples incorrectly updated. please fix it.

File: bin/configs/go-petstore.yaml [hunk 2]
```diff
@@ -15,6 +15,7 @@ additionalProperties:
   packageName: petstore
   disallowAdditionalPropertiesIfNotPresent: false
   generateInterfaces: true
+  initRequiredVars: true
```
Comment (wing328) [inline bin/configs/go-petstore.yaml:18]: please rename this option to "useDefaultValueInRequiredVars" instead

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache [hunk 3]
```diff
@@ -168,6 +168,18 @@ func (o *{{classname}}) Set{{name}}(v {{vendorExtensions.x-go-base-type}}) {
...
+{{#defaultValue}}
+{{#initRequiredVars}}
+func (o *{{classname}}) GetDefault{{baseName}}() interface{}  {
+	var {{baseName}} {{{dataType}}} = {{{defaultValue}}}
+	o.{{name}} = {{baseName}}
+	return {{baseName}}
+}
+{{/initRequiredVars}}
+{{/defaultValue}}
+{{/isReadOnly}}
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/model_simple.mustache:181]: Better wrap the whole block inside `{{#useDefaultValueInRequiredVars}} ... {{/useDefaultValueInRequiredVars}}` instead

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache [hunk 4]
```diff
@@ -168,6 +168,18 @@ func (o *{{classname}}) Set{{name}}(v {{vendorExtensions.x-go-base-type}}) {
 {{/isNullable}}
 }
 
+{{^isReadOnly}}
+{{#defaultValue}}
+{{#initRequiredVars}}
+func (o *{{classname}}) GetDefault{{baseName}}() interface{}  {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/model_simple.mustache:174]: what about adding a docstring (a line or 2) explaining what this function does?

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache [hunk 5]
```diff
@@ -168,6 +168,18 @@ func (o *{{classname}}) Set{{name}}(v {{vendorExtensions.x-go-base-type}}) {
...
+	var {{baseName}} {{{dataType}}} = {{{defaultValue}}}
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/model_simple.mustache:175]: i don't think we need an additional variable to hold the default value which can be used directly in the line below

File: samples/client/echo_api/php-nextgen/.openapi-generator/FILES [hunk 6]
```diff
@@ -24,13 +24,13 @@ docs/Model/TestQueryStyleDeepObjectExplodeTrueObjectAllOfQueryObjectParameter.md
...
 git_push.sh
 phpunit.xml.dist
+src/ApiException.php
 src/Api/AuthApi.php
 src/Api/BodyApi.php
 src/Api/FormApi.php
 src/Api/HeaderApi.php
 src/Api/PathApi.php
 src/Api/QueryApi.php
-src/ApiException.php
```
Comment (wing328) [inline samples/client/echo_api/php-nextgen/.openapi-generator/FILES:33]: are you using git bash or WSL to run the sample update script under windows?

Comment (prmedipa) [reply samples/client/echo_api/php-nextgen/.openapi-generator/FILES:33]: I am using git bash

Comment (wing328) [reply samples/client/echo_api/php-nextgen/.openapi-generator/FILES:33]: can you try WSL instead?

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache (hunk 4 above)
Comment (prmedipa) [reply modules/openapi-generator/src/main/resources/go/model_simple.mustache]: I have addressed

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache (hunk 3 above)
Comment (prmedipa) [reply modules/openapi-generator/src/main/resources/go/model_simple.mustache:181]: I have addressed

File: bin/configs/go-petstore.yaml (hunk 2 above)
Comment (prmedipa) [reply bin/configs/go-petstore.yaml]: addressed

**Review by vvb — CHANGES_REQUESTED** [review]
@prmedipa Run the cmds in a unix/linux environment and update the PR. There are a lot of unnecessary diff because of the windows related tooling.

**Comment by prmedipa** [comment]
@vvb @wing328 
Please review the following pull request: https://github.com/OpenAPITools/openapi-generator/pull/19218.
The file samples/openapi3/client/petstore/go/go-petstore/model_category.go contains code for default values.

**Pull request closed without merging**